ports/nrf/common-hal/bleio/Peripheral.c

```diff
@@ -0,0 +1,334 @@
+#define BLE_SLAVE_LATENCY            0
+#define BLE_CONN_SUP_TIMEOUT         MSEC_TO_UNITS(4000, UNIT_10_MS)
+
+#define BLE_ADV_LENGTH_FIELD_SIZE   1
+#define BLE_ADV_AD_TYPE_FIELD_SIZE  1
+#define BLE_AD_TYPE_FLAGS_DATA_SIZE 1
+
+static uint8_t m_adv_handle = BLE_GAP_ADV_SET_HANDLE_NOT_SET;
+
+STATIC void check_data_fit(size_t pos, size_t data_len) {
+    if (pos + data_len >= BLE_GAP_ADV_SET_DATA_SIZE_MAX) {
+        mp_raise_ValueError(translate("Data too large for advertisement packet"));
+    }
+}
+
+STATIC uint32_t add_services_to_advertisement(bleio_peripheral_obj_t *self, size_t* adv_data_pos_p, size_t uuid_len) {
+    uint32_t uuids_total_size = 0;
+    const mp_obj_list_t *service_list = MP_OBJ_TO_PTR(self->service_list);
+    uint32_t err_code = NRF_SUCCESS;
+
+    check_data_fit(*adv_data_pos_p, 1 + 1);
+
+    // Remember where length byte is; fill in later when we know the size.
+    const size_t length_pos = *adv_data_pos_p;
+    (*adv_data_pos_p)++;
+
+    self->adv_data[(*adv_data_pos_p)++] = (uuid_len == 16)
+        ? BLE_GAP_AD_TYPE_16BIT_SERVICE_UUID_COMPLETE
+        : BLE_GAP_AD_TYPE_128BIT_SERVICE_UUID_COMPLETE;
+
+    for (size_t i = 0; i < service_list->len; ++i) {
+        const bleio_service_obj_t *service = MP_OBJ_TO_PTR(service_list->items[i]);
+        uint8_t encoded_size = 0;
+
+        // Skip services of the wrong length and secondary services.
+        if (common_hal_bleio_uuid_get_size(service->uuid) != uuid_len || service->is_secondary) {
+            continue;
+        }
+
+        ble_uuid_t uuid;
+        bleio_uuid_convert_to_nrf_ble_uuid(service->uuid, &uuid);
+
+        err_code = sd_ble_uuid_encode(&uuid, &encoded_size, &(self->adv_data[*adv_data_pos_p]));
+        if (err_code != NRF_SUCCESS) {
+            return err_code;
+        }
+
+        check_data_fit(*adv_data_pos_p, encoded_size);
+        uuids_total_size += encoded_size;
+        (*adv_data_pos_p) += encoded_size;
+    }
+
+    self->adv_data[length_pos] = 1 + uuids_total_size;  // 1 for the field type.
+    return err_code;
+}
+
+
+
+STATIC uint32_t set_advertisement_data(bleio_peripheral_obj_t *self, bool connectable, mp_buffer_info_t *raw_data) {
+    common_hal_bleio_adapter_set_enabled(true);
+
+    size_t adv_data_pos = 0;
+    uint32_t err_code;
+
+    GET_STR_DATA_LEN(self->name, name_data, name_len);
+    if (name_len > 0) {
+        ble_gap_conn_sec_mode_t sec_mode;
+        BLE_GAP_CONN_SEC_MODE_SET_OPEN(&sec_mode);
+
+        // We'll add the name after everything else, shortening it if necessary.
+        err_code = sd_ble_gap_device_name_set(&sec_mode, name_data, name_len);
+        if (err_code != NRF_SUCCESS) {
+            return err_code;
+        }
+    }
+
+    if (raw_data->len != 0) {
+        // User-supplied advertising packet.
+        check_data_fit(adv_data_pos, raw_data->len);
+        memcpy(&(self->adv_data[adv_data_pos]), raw_data->buf, raw_data->len);
+        adv_data_pos += raw_data->len;
+    } else {
```

Comment:
Setting an empty advertising packet seems weird. Should the constructor just make a default set of data?

Reply:
Not sure I understand you. If you don't supply a packet (which is the highly normal case), it will create a default advertising packet starting at line 127.

The code for the default adveritisement may eventually move into `AdvertisingData.py` and `adafruit_ble.PeripheralServer` (or whatever i decide to call it). The `bleio.Peripheral` class will then always take a raw packet and will no longer be for use by mere mortals.

Reply:
Oh, I see what you're asking. It's just that that `shared-bindings/bleio/Peripheral.c` will pass down an empty `raw_data` buffer if the user supplies `None` to the advertisement data packet.

Reply:
Ok, added a request for more comments.

Reply:
Added comments explaining `data=None` to both shared-bindings and common-hal.